### PR TITLE
Add exists check for workflow & node executions

### PR DESCRIPTION
**Pull request by katrogan**
Signed-off-by: Katrina Rogan <katroganGH@gmail.com>

# TL;DR
Add exists check for workflow & node executions to avoid fetching and passing around fully fledged executions in memory.

Additionally, only add the exists check for these parent executions when node and task executions, respectively, are not found and we create an entry in the db for them for the first time.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Tracking Issue
https://github.com/flyteorg/flyte/issues/880

## Follow-up issue
_NA_

### CODE REVIEW

**Review by kumare3 — APPROVED** [review]
This is great!

**Comment by kumare3** [comment]
I tested this change, seems good to go, performance improvement is not great, but this is the right change nevertheless
